core/shared/src/main/scala/zio/stm/TPriorityQueue.scala

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2017-2020 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.stm
+
+import scala.collection.immutable.SortedMap
+
+/**
+ * A simple `TPriorityQueue` implementation. A `TPriorityQueue` contains values
+ * of type `V`. Each value is associated with a key of type `K` that an
+ * `Ordering` is defined on. Unlike a `TQueue`, `take` returns the highest
+ * priority value (the value that is first in the specified ordering) as
+ * opposed to the first value offered to the queue. The ordering that elements
+ * with the same priority will be taken from the queue is not guaranteed.
+ */
+final class TPriorityQueue[K, V] private (private val tref: TRef[SortedMap[K, ::[V]]]) extends AnyVal {
+
+  /**
+   * Offers the specified value to the queue with the specified priority.
+   */
+  def offer(key: K, value: V): USTM[Unit] =
+    tref.update { map =>
+      map.get(key) match {
+        case None         => map + (key -> ::(value, Nil))
+        case Some(values) => map + (key -> ::(value, values))
+      }
+    }
+
+  /**
+   * Offers all of the elements in the specified collection to the queue.
+   */
+  def offerAll(values: Iterable[(K, V)]): USTM[Unit] =
+    ZSTM.foreach_(values) { case (key, value) => offer(key, value) }
+
+  /**
+   * Peeks at the first value in the queue without removing it, retrying until
+   * a value is in the queue.
+   */
+  def peek: USTM[V] =
+    tref.get.flatMap { map =>
+      map.headOption match {
+        case None                  => ZSTM.retry
```

Comment:
with ZSTM, we have have the advantage of not blocking here, yes?

Reply:
@phderome yup,suspended until changes occur.